darts/models/forecasting/tft_model.py

```diff
@@ -0,0 +1,915 @@
+
+
+class TFTModel(TorchParametricProbabilisticForecastingModel, MixedCovariatesTorchModel):
+    @random_method
+    def __init__(self,
+                 input_chunk_length: int = 12,
+                 output_chunk_length: int = 1,
+                 hidden_size: Union[int, List[int]] = 16,
+                 lstm_layers: int = 1,
+                 num_attention_heads: int = 4,
+                 dropout: float = 0.1,
+                 hidden_continuous_size: int = 8,
+                 add_cyclic_encoder: Optional[str] = None,
+                 loss_fn: Optional[nn.Module] = None,
+                 likelihood: Optional[Likelihood] = None,
+                 max_samples_per_ts: Optional[int] = None,
+                 random_state: Optional[Union[int, RandomState]] = None,
+                 **kwargs
+                 ):
+        """Temporal Fusion Transformers (TFT) for Interpretable Time Series Forecasting.
+
+        This is an implementation of the TFT architecture, as outlined in this paper:
+        https://arxiv.org/pdf/1912.09363.pdf.
+
+        The internal sub models were adopted from `pytorch-forecasting's TemporalFusionTransformer
+        <https://pytorch-forecasting.readthedocs.io/en/latest/models.html>`_ implementation.
+
+        This model supports mixed covariates (includes past covariates known for `input_chunk_length`
+        points before prediction time and future covariates known for `output_chunk_length` after prediction time).
+
+        The TFT applies multi-head attention queries on future inputs from mandatory `future_covariates`.
+        Specifying `add_cyclic_encoder` (read below) adds cyclic temporal encoding to the model and allows to use
+        the model without having to specify additional `future_covariates` for training and prediction.
+
+        By default, this model uses the ``QuantileRegression`` likelihood, which means that its forecasts are
+        probabilistic; it is recommended to call ``predict()`` with ``num_samples >> 1`` to get meaningful results.
+
+        Parameters
+        ----------
+        input_chunk_length : int
+            encoder length; number of past time steps that are fed to the forecasting module at prediction time.
+        output_chunk_length : int
+            decoder length; number of future time steps that are fed to the forecasting module at prediction time.
+        hidden_size : int
+            hidden state size of the TFT. It is the main hyper-parameter and common across the internal TFT architecture.
+        lstm_layers : int
+            number of layers for the Long Short Term Memory (LSTM) Encoder and Decoder (2 is a good default).
+        num_attention_heads : int
+            number of attention heads (4 is a good default)
+        dropout : float
+            Fraction of neurons afected by Dropout.
+        hidden_continuous_size : int
+            default for hidden size for processing continuous variables
+        add_cyclic_encoder : optional str
```

Comment:
Could we find a way to make this one work for integer-indexed time series?
I think in such cases, we could "simply" generate some sine-based encodings of the positions; similar to what is done in the original transformer paper. I guess we could perhaps simply change the `datetime_attribute_timeseries()` to handle those integer-series cases too. Maybe dynamically ignoring the `attribute` argument when `time_index` is a `pd.Int64Index`.

If that is too complex, it can be left out for a future PR :) But it'd be nice to have at some point.

Reply:
as discussed with @hrzn, we should do this in another PR. Opened ticket [DARTS-499](https://unit8.atlassian.net/browse/DARTS-499)